### ISSUE TITLE
DockerV2 using ACR - no such host

### ISSUE THREAD

**Issue by adam-resdiary** [microsoft/azure-pipelines-tasks#10616]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: DockerV2

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - Azure Pipelines
- Agent - Hosted ubuntu-latest

## Issue Description

We're trying to build and push a docker image to a private Azure Container Registry. The Docker Login step succeeds, but the Docker buildAndPush step fails with the following error message:

```shell
Get https://***/v2/: dial tcp: lookup *** on 168.63.129.16:53: no such host
/usr/bin/docker failed with return code: 1
```

This sounds like it might be related to https://github.com/microsoft/azure-pipelines-tasks/issues/10612, although we're getting a slightly different error message and using a different agent.

Here is a stripped down version of our pipeline yaml:

```yaml
name: "2019.3$(Rev:.r)"

trigger:
- master

pool:
  vmImage: 'ubuntu-latest'

variables:
  configuration: Release
  fullVersion: "2019.3.$(Build.BuildId)"

steps:
- task: Docker@2
  displayName: Login to ACR
  inputs:
    command: login
    containerRegistry: resdiary-acr

- task: Docker@2
  displayName: Build and push docker image
  inputs:
    command: buildAndPush
    containerRegistry: resdiary-acr
    repository: resdiary/database-maintenance
    tags: $(fullVersion)
```

### Task logs

```shell
##[section]Starting: Build and push docker image
==============================================================================
Task         : Docker
Description  : Build or push Docker images, login or logout, or run a Docker command
Version      : 2.152.3
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/build/docker
==============================================================================
[command]/usr/bin/docker build -f /home/vsts/work/1/s/Dockerfile --label com.azure.dev.image.system.teamfoundationcollectionuri=https://dev.azure.com/xyz/ --label com.azure.dev.image.system.teamproject=Database Maintenance --label com.azure.dev.image.build.repository.name=xyz --label com.azure.dev.image.build.repository.uri=https://github.com/xyz --label com.azure.dev.image.build.sourcebranchname=merge --label com.azure.dev.image.build.sourceversion=28fd3aa0e53b82028e8247958e993c39f0aa10f3 --label com.azure.dev.image.build.definitionname=xyz --label com.azure.dev.image.build.buildnumber=2019.3.14 --label com.azure.dev.image.build.builduri=vstfs:///Build/Build/27 --label com.azure.dev.image.build.requestedfor=GitHub -t ***/resdiary/database-maintenance:2019.3.27 /home/vsts/work/1/s
Sending build context to Docker daemon  5.492MB

Step 1/15 : FROM mcr.microsoft.com/dotnet/core/runtime:2.2
2.2: Pulling from dotnet/core/runtime
743f2d6c1f65: Pulling fs layer
074da88b8de0: Pulling fs layer
ac831735b47a: Pulling fs layer
3adcc844418d: Pulling fs layer
3adcc844418d: Waiting
ac831735b47a: Download complete
074da88b8de0: Verifying Checksum
074da88b8de0: Download complete
3adcc844418d: Verifying Checksum
3adcc844418d: Download complete
743f2d6c1f65: Verifying Checksum
743f2d6c1f65: Download complete
743f2d6c1f65: Pull complete
074da88b8de0: Pull complete
ac831735b47a: Pull complete
3adcc844418d: Pull complete
Digest: sha256:d9543cbaa0d6ad7e0cda582608cde22da424c932e7d17ecdc433da0c1a8ceabb
Status: Downloaded newer image for mcr.microsoft.com/dotnet/core/runtime:2.2
 ---> c0f9ab44ecc1
Step 2/15 : WORKDIR /app
 ---> Running in 17b4ee6f68d8
Removing intermediate container 17b4ee6f68d8
 ---> 751998d2ad19
Step 3/15 : COPY artifacts/output .
 ---> 7820d2519b7d
Step 4/15 : COPY src/appsettings.yaml .
 ---> ddf985fc1521
Step 5/15 : ENTRYPOINT ["dotnet", "RD.DatabaseMaintenance.dll"]
 ---> Running in aa825ab58117
Removing intermediate container aa825ab58117
 ---> e38d38095bcd
Step 6/15 : LABEL com.azure.dev.image.build.buildnumber=2019.3.14
 ---> Running in df252650997f
Removing intermediate container df252650997f
 ---> cd9c1cdde28c
Step 7/15 : LABEL com.azure.dev.image.build.builduri=vstfs:///Build/Build/27
 ---> Running in d68e84c5a0bb
Removing intermediate container d68e84c5a0bb
 ---> 503e70720965
Step 8/15 : LABEL com.azure.dev.image.build.definitionname=xyz
 ---> Running in a5fc5b334f54
Removing intermediate container a5fc5b334f54
 ---> a2f74a7ccc92
Step 9/15 : LABEL com.azure.dev.image.build.repository.name=xyz
 ---> Running in 037dba944eed
Removing intermediate container 037dba944eed
 ---> 5af630af58fc
Step 10/15 : LABEL com.azure.dev.image.build.repository.uri=https://github.com/xyz
 ---> Running in 5a7debc70dba
Removing intermediate container 5a7debc70dba
 ---> f7a1b281915a
Step 11/15 : LABEL com.azure.dev.image.build.requestedfor=GitHub
 ---> Running in b2096844c03b
Removing intermediate container b2096844c03b
 ---> 70bfcdbe86b0
Step 12/15 : LABEL com.azure.dev.image.build.sourcebranchname=merge
 ---> Running in 41c144d0fdbc
Removing intermediate container 41c144d0fdbc
 ---> adb986725dce
Step 13/15 : LABEL com.azure.dev.image.build.sourceversion=28fd3aa0e53b82028e8247958e993c39f0aa10f3
 ---> Running in 3e542ba7de7e
Removing intermediate container 3e542ba7de7e
 ---> 7387addddad9
Step 14/15 : LABEL com.azure.dev.image.system.teamfoundationcollectionuri=https://dev.azure.com/xyz/
 ---> Running in db5057b2ee69
Removing intermediate container db5057b2ee69
 ---> a64b38582c25
Step 15/15 : LABEL com.azure.dev.image.system.teamproject=Database Maintenance
 ---> Running in c925ec9f8ed6
Removing intermediate container c925ec9f8ed6
 ---> 82181ab24445
Successfully built 82181ab24445
Successfully tagged ***/resdiary/database-maintenance:2019.3.27
[command]/usr/bin/docker push ***/resdiary/database-maintenance:2019.3.27
The push refers to repository [***/resdiary/database-maintenance]
Get https://***/v2/: dial tcp: lookup *** on 168.63.129.16:53: no such host
##[error]Get https://***/v2/: dial tcp: lookup *** on 168.63.129.16:53: no such host
##[error]/usr/bin/docker failed with return code: 1
##[section]Finishing: Build and push docker image
```

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs

```shell
Get https://***/v2/: dial tcp: lookup *** on 168.63.129.16:53: no such host
/usr/bin/docker failed with return code: 1
```

**Comment by bansalaseem** [comment]
@adam-resdiary Are you still running into this issue?

**Comment by adam-resdiary** [comment]
@bansalaseem I tried this again just after #10612 was resolved, and it still wasn't working. If you think it might have been fixed since then I'll try again when I've got a chance.

**Comment by ajinkya599** [comment]
@adam-resdiary - Will need your help with the following:

Can you please try it again and let us know ?
Are you able to push to the same registry from your local box ?
Is the registry restricted to some VNet or configured behind a firewall ?

**Comment by adam-resdiary** [comment]
@ajinkya599 I've just tried it again with exactly the same result. Here's a snippet of the log:

```text
Successfully tagged ***/resdiary/database-maintenance-actest:1.0.0-152 |  
-- | --
  | sr/bin/docker push ***/resdiary/database-maintenance-actest:1.0.0-152 |  
  | The push refers to repository [***/resdiary/database-maintenance-actest] |  
  | Get https://***/v2/: dial tcp: lookup *** on 168.63.129.16:53: no such host |  
  | ##[error]Get https://***/v2/: dial tcp: lookup *** on 168.63.129.16:53: no such host |  
  | ##[error]The process '/usr/bin/docker' failed with exit code 1 |  
  | Finishing: Build and push docker image
```

To answer your other questions:

- The registry I'm trying to push to is a private Azure Container Registry.
- I can push to it from my local box after running the `az acr login` command.
- The registry is not on a VNet and has no firewall restrictions.

I'm not sure if this is relevant or not, but the registry was created quite a long time ago, and has a tenant suffix (i.e. the registry name is something like `myregistry-on` instead of just `myregistry`). I can login fine without specifying the suffix using `az acr login`.

**Comment by ajinkya599** [comment]
> I'm not sure if this is relevant or not, but the registry was created quite a long time ago, and has a tenant suffix (i.e. the registry name is something like `myregistry-on` instead of just `myregistry`). I can login fine without specifying the suffix using `az acr login`.

This can be relevant.
- What is the `registry url` ? (You can get it from `Azure portal -> Your registry -> Access keys -> Login server`). 
- Can you find out what `url` is stored in the `auths` section in your docker config file (in .docker/config.json) on your local box after running `az acr login`?
- Is the `registry / loginserver` in the service connection identical to them ?
- Which registry type are you using in the service connection ?
![image](https://user-images.githubusercontent.com/11447401/62459122-33011180-b79c-11e9-9ba9-78a6398b23ac.png)

Please share the above information with us on mail to RM_Customer_Queries at the rate of microsoft dot com.

About the tenant suffix part:
- Do you mean it is something like `myregistry-mytenant` ? I see at this point ACR does not allow `-` in registry.
![image](https://user-images.githubusercontent.com/11447401/62458657-1ca68600-b79b-11e9-85bb-de7bca00d0ee.png)
- Are you able to login with the name `myregistry` which is not the complete name ?
- Are you able to login with the suffix as well ?

**Comment by adam-resdiary** [comment]
@ajinkya599 I've just emailed the answers to your questions. Sorry for the delayed reaction - this has just been less of a priority since we worked around it using the script task.

I can login to the registry using either just the registry name, or the registry name and tenant suffix. For example:

```shell
az acr login --name myregistry
az acr login --name myregistry --suffix on
```

The second command logs in fine, but returns the following warning:

```shell
WARNING: Obtained registry login server 'myregistry-on.azurecr.io' from service. The specified suffix 'on' is ignored.
Login Succeeded
```

So just to clarify, the "on" part isn't part of the registry name, it was just automatically appended to the login server name when the registry was created.

**Comment by ajinkya599** [comment]
@adam-resdiary - There was an issue with the service endpoint creation some time back. Can you try creating a new endpoint and use it in the pipeline ?

**Comment by ajinkya599** [comment]
@adam-resdiary - Were you able to try this out ?

**Comment by ajinkya599** [comment]
Closing due to inactivity. Please feel free to re-open the issue if its still there.

**Comment by adam-resdiary** [comment]
@ajinkya599 no problem. Sorry, but it's just not been a big priority since we were able to use the script task fine. If we get a chance to re-test it I'll update this, but it sounds like it's probably not an issue anymore from what you're saying.

**Comment by JimPaine** [comment]
Currently hitting this with a Service Connection and Environment that is only a week old, using ACR for service connection. Is this something your still aware of being an issue?

**Comment by szwed** [comment]
Same issue here: 
```
2020-04-08T10:10:48.2673484Z Error response from daemon: Get https://***/v2/: dial tcp: lookup *** on 127.0.0.53:53: no such host
```

**Comment by ponnmohan** [comment]
>  s

Any solution? issue fixed?

**Comment by philipwolfe** [comment]
@ponnmohan, @deepaksinghnationwidecouk, @szwed, @ajinkya599 and to anyone else 
I found that my AWS URL was wrong.  I found that mine was `...amazonws.com` instead of '...amazonaws.com`.  I was missing the 'A' in 'AWS' and I received this error:
```
The push refers to repository [000000000000.dkr.ecr.us-east-2.amazonws.com/review-apps]
Get https://000000000000.dkr.ecr.us-east-2.amazonws.com/v2/: dial tcp: lookup 000000000000.dkr.ecr.us-east-2.amazonws.com: no such host
```

**Comment by sekhemrekhutawysobekhotep** [comment]
this issue is ongoing ... I am getting

Get "https://***/v2/": dial tcp: lookup ***: no such host


```

Removing intermediate container f8c98b9a0cd8
 ---> 84cb8945076e
Step 18/18 : LABEL image.base.ref.name=node:10
 ---> Running in 1f0b56e8cc00
Removing intermediate container 1f0b56e8cc00
 ---> 4c3c44bd8cb6
Successfully built 4c3c44bd8cb6
Successfully tagged ***/sehotetoynodeap00:1408
/usr/bin/docker images
/usr/bin/docker push ***/sehotetoynodeap00:1408
REPOSITORY                                                   TAG         IMAGE ID       CREATED         SIZE
***/sehotetoynodeap00   1408        4c3c44bd8cb6   1 second ago    914MB
alpine                                                       3.13        38cacb9bafd2   2 weeks ago     5.61MB
alpine                                                       3.14        5977be310a9d   2 weeks ago     5.59MB
node                                                         14-alpine   937e8a02e5c7   2 weeks ago     119MB
node                                                         16-alpine   b0cbdedc1b9d   2 weeks ago     112MB
node                                                         14          326f034bd14b   3 weeks ago     914MB
node                                                         16          c6b745e900c7   3 weeks ago     907MB
buildpack-deps                                               buster      61e83b993b64   3 weeks ago     804MB
buildpack-deps                                               bullseye    73b7450f22be   3 weeks ago     834MB
debian                                                       10          61cd0373ceab   3 weeks ago     114MB
debian                                                       11          123c2f3835fd   3 weeks ago     124MB
buildpack-deps                                               stretch     9d4f40bcbdd3   5 weeks ago     835MB
debian                                                       9           662c05203bab   5 weeks ago     101MB
ubuntu                                                       20.04       20fffa419e3a   8 weeks ago     72.8MB
ubuntu                                                       18.04       ad080923604a   8 weeks ago     63.1MB
moby/buildkit                                                latest      a2c9241854f2   2 months ago    142MB
node                                                         12          6c8de432fc7f   3 months ago    918MB
node                                                         12-alpine   bb6d28039b8c   3 months ago    91MB
alpine                                                       3.12        24c8ece58a1a   4 months ago    5.58MB
ubuntu                                                       16.04       b6f507652425   11 months ago   135MB
node                                                         10          28dca6642db8   16 months ago   910MB
The push refers to repository [***/sehotetoynodeap00]
Get "https://***/v2/": dial tcp: lookup ***: no such host
##[error]Get "https://***/v2/": dial tcp: lookup ***: no such host
##[error]The process '/usr/bin/docker' failed with exit code 1
Finishing: Build and push an image to container registry



```

here is the azure DevOps pipeline

```
stages:
- stage: Build
  displayName: Build and push stage
  jobs:
  - job: Build
    displayName: Build
    pool:
      vmImage: $(vmImageName)
    steps:
    - task: Docker@2
      displayName: Build and push an image to container registry
      inputs:
        command: buildAndPush
        repository: $(imageRepository)
        dockerfile: $(dockerfilePath)
        containerRegistry: $(dockerRegistryServiceConnection)
        tags: |
          $(tag)



```

above is first step of pipeline and above DID run OK then after several re-runs above pipeline failed with above error

**Comment by Teketel** [comment]
I am also getting this issue. Do we find a solution? Please help me!
https://github.com/microsoft/azure-pipelines-tasks/issues/10616#issuecomment-1204409194

**Comment by ajayess** [comment]
I recently saw this error when I used Docker@2 command for pushing images to Azure Container Registry.

![Image](https://github.com/user-attachments/assets/6e608138-a7e0-46dc-95c0-848d21d2e5ce)

Previously, we were using Service Connection of Azure subscription. It was working well with Docker@1 task. However, Docker@2 task requires service connection of ACR.

![Image](https://github.com/user-attachments/assets/3fe3b526-66d3-4f6f-aec7-7300232d33db)

using this ACR service connection fix the error for me in Azure Build Pipeline.
